### PR TITLE
[IMP]better defensive on export strategy.

### PR DESCRIPTION
**Pull request by flotho**

### CODE REVIEW

File: connector_magento_custom_attributes/models/product/listener.py
Comment (wpichler) [inline connector_magento_custom_attributes/models/product/listener.py]: This check should not be here - it should be in the base Modul - in connector_magento_product_catalog

File: connector_magento_product_catalog/models/product/listener.py
Comment (wpichler) [inline connector_magento_product_catalog/models/product/listener.py]: We should generally add a force context option here - so that the export is possible also if magento_first is set. For example for the Direct Export Button on the binding record. If we don't add it - then we should disable the export buttons on the binding records - because they would have no affect...

**Review by wpichler — CHANGES_REQUESTED** [review]

File: connector_magento_product_catalog/models/product/listener.py
Comment (flotho) [reply connector_magento_product_catalog/models/product/listener.py]: Ok,

I agree that we should express more clearly the available options, Yet I'm not sure that our immediate proposals are accurate. 

What i would like to clearly present for the user is the fact that its backend is either set for import (classical connector_magento module) or for export (first option + all the catalogue functions).

If you share my point of view I would propose to hide the export / Import button regarding the backend configuration.

Let me know

Comment (wpichler) [reply connector_magento_product_catalog/models/product/listener.py]: I think the original concept was an other concept...


The exporter function does call the _should_import function - this function does check if the odoo last modification date is greater than the magento last modification date. If the magento date is greater - then it will return true.

If it does return true - then the _delay_import will get called. In the _delay_import now we should check the sync strategy - if it is set to magento_first - then run immediate the import - and then the export. If set to odoo first - then run the export first - and delay the import.

So theoretically it should never check the sync strategy in the listener functions - because the export job should get run - and the export job then does the check.

Comment (flotho) [reply connector_magento_product_catalog/models/product/listener.py]: ok, I understand and I also remember.

Comment (flotho) [reply connector_magento_product_catalog/models/product/listener.py]: @wpichler , could you just remind me why the _should_import method is named this way as it is only used in export ?

Comment (wpichler) [reply connector_magento_product_catalog/models/product/listener.py]: the _should_import function should decide if the record should get imported before it gets exported. If _should_import does return true - then the record on magento side does has a change that we don't have on odoo side. With the selected strategy we have to decide if we do import the record before the export (so magento data does overwrite odoo data = magento_first) or if we do export the record - and then import the record from magento (=odoo_first)

Comment (wpichler) [reply connector_magento_product_catalog/models/product/listener.py]: As far as i can see in the code - is it currently more a mess than consistent implemented. So this can get refactored completly

Comment (flotho) [reply connector_magento_product_catalog/models/product/listener.py]: ok, thanks

Comment (flotho) [reply connector_magento_product_catalog/models/product/listener.py]: BTW, maybe the field strategy is not properly named.
My idea was to either import from magento product data or export product data. 
Did you understood this field the same way? Do I need to refactor as well ?

**Comment by flotho** [comment]
Hi @wpichler , we have additional question.
Is it normal that you bind a stock.item to a configurable ther : https://github.com/wpichler/connector-magento/blob/10.0-callino-rebase/connector_magento/models/stock_item/importer.py#L42
IMHO, stock is only set on simpel product in magento as well as product.product in Odoo.
Do you agree ?

File: connector_magento_product_catalog/models/product/listener.py
Comment (wpichler) [reply connector_magento_product_catalog/models/product/listener.py]: I did understand the field as priority field - it does specify if magento data has priority over odoo data or the other way round. A field for export or not is not needed - because the export functionality is in a separate module - so if you don't install export module - then you want have the export. I think we should refactor it to be clearly a priority strategy field.

**Review by wpichler — CHANGES_REQUESTED** [review]
_get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values

**Comment by flotho** [comment]
> _get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values

In the same way, won't it be coherent to migrate exporter and listener of magento.attribute from the base module to the catalog one ?

File: connector_magento/models/product/importer.py
Comment (wpichler) [inline connector_magento/models/product/importer.py:198]: As far as i know - the weight in odoo is fixed to kg - it does not depend on the product uom. So weight in magento and odoo is in kg - no need for this in my opinion.

**Comment by wpichler** [comment]
> > _get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values
> 
> In the same way, won't it be coherent to migrate exporter and listener of magento.attribute from the base module to the catalog one ?

this is a good question...

In the base module there must be everything to get a working magento to odoo import. There is no direct need for the attribute export functions as far as i can tell.

For me it would be ok to migrate the attributes to catalog module.

**Comment by flotho** [comment]
@wpichler , additionnal question regarding the dropoff_site https://github.com/wpichler/connector-magento/tree/10.0-callino-rebase/connector_magento_dropoffsite

Which magento module are you using ?

Regards

**Comment by wpichler** [comment]
> @wpichler , additionnal question regarding the dropoff_site https://github.com/wpichler/connector-magento/tree/10.0-callino-rebase/connector_magento_dropoffsite
> 
> Which magento module are you using ?
> 
> Regards

I've never worked on the dropoffsite module

File: connector_magento/models/product/importer.py
Comment (wpichler) [reply connector_magento/models/product/importer.py:198]: I think we can leave this change in there - because maybe sometimes magento will change the default weight - and then we are prepared...
Setting the default to kg would be great.

**Comment by wpichler** [comment]
Hi - i'd like to merge this - are you done with your main changes ?

**Comment by flotho** [comment]
Hi @wpichler 

Best wishes for this new year
I didn't make all the refactoring we talked about but you could merge it

**Comment by flotho** [comment]
ping @wpichler ,

Any chance to be merge as there is no additionnal dev from my side ?

File: connector_magento/models/product_template/importer.py
Comment (wpichler) [inline connector_magento/models/product_template/importer.py:253]: This is a problem here - this member is already in use in the magento.product.product.import.mapper - from which this does inherit. So you should not overwrite it - just extend it.
I will correct this with a new commit after the merge

Comment (flotho) [reply connector_magento/models/product_template/importer.py:253]: Great thanks

**Comment by flotho** [comment]
thanks for accepting this job.
I really appreciate to have a partner on this.

Regards

**Comment by wpichler** [comment]
No problem. Also many thank for your hard work. This is such a big project - it would require more dev power ;-)

What do you think about pushing this into odoo app store ?
I would also love to see this in main oca repositories - but i think we will fail on beautifying the code - and adding test routines...

**Pull request merged**
